### PR TITLE
[docs] Adds GRPO, RSO and LoRA to Paper Index

### PR DESCRIPTION
**Pull request by SSusantAchary**
## Summary
Expands the Paper Index with 10 additional papers spanning GRPO/DPO/CPO, SFT/PEFT, and foundational systems. Entries follow the house style (title → 📜 link → one-line summary → trainer cross-link when relevant).

### Group Relative Policy Optimization
- **DeepSeekMath: Pushing the Limits of Mathematical Reasoning in Open Language Models** (2402.03300) — Introduces GRPO and shows strong math-reasoning gains.  
  **Used in TRL via:** [`GRPOTrainer`]

### Direct Policy Optimization
- **Statistical Rejection Sampling Improves Preference Optimization** (2309.06657) — RSO for better preference pairs; complements DPO/SLiC.
- **Nash Learning from Human Feedback** (2312.00886) — Frames alignment as a two-player game with Nash policies.
- **Direct Language Model Alignment from Online AI Feedback** (2402.04792) — Online AI feedback signals for direct alignment.

### Contrastive Preference Optimization
- **Contrastive Preference Optimization: Pushing the Boundaries of LLM Performance in Machine Translation** (2401.08417) — Contrastive pairs to avoid adequate-but-suboptimal outputs.  
  **Used in TRL via:** [`CPOTrainer`]

### Supervised Fine-Tuning
- **LoRA: Low-Rank Adaptation of Large Language Models** (2106.09685) — Parameter-efficient adapters; reference for TRL PEFT integration.

### Reward / Process Supervision (Background)
- **Solving Math Word Problems with Process- and Outcome-Based Feedback** (2211.14275) — Motivates process supervision signals.

### Distillation / Post-training (Background)
- **On-Policy Distillation of Language Models** (2306.13649) — GKD on-policy student/teacher distillation.

### Foundations & Systems (Background)
- **Proximal Policy Optimization Algorithms** (1707.06347) — PPO foundation used across RL/RLHF variants.
- **ZeRO: Memory Optimizations Toward Training Trillion-Parameter Models** (1910.02054) — Partitioning states/grad/params for scale.

## Implementation details
- Kept headings/emoji/spacing consistent with existing sections.
- Added/confirmed link-refs:
  - [`GRPOTrainer`]: `grpo_trainer`
  - [`CPOTrainer`]: `cpo_trainer`
- Avoided duplicates (scanned current page before insertion).

## Scope
- Docs only (`docs/source/paper_index.md`).
- No API or code changes.

## Checklist
- [x] Neutral one-liners (≤ ~30 words)
- [x] Live links (HF paper pages + trainer docs where applicable)
- [x] Consistent style with existing entries
- [x] References to **#4407** included

Relates to **#4407**.

### CODE REVIEW

File: docs/source/paper_index.md [hunk 1]
```diff
@@ -231,6 +231,11 @@ trainer = PAPOTrainer(
     ...
 )
 ```
+### DeepSeekMath: Pushing the Limits of Mathematical Reasoning in Open Language Models
+**📜 Paper**: https://huggingface.co/papers/2402.03300
+
+Introduces **GRPO** and shows strong math-reasoning gains from math-centric pretraining plus group-relative PPO-style optimization.  
+**Used in TRL via:** [`GRPOTrainer`]
```
Comment (qgallouedec) [inline docs/source/paper_index.md]: Since this paper introduced GRPO, I recommend having it first in this section. Also, it would be neat to have a code snippet that gives the hyperparameters mentioned in the papers, as with the other papers.

File: docs/source/paper_index.md [hunk 2]
```diff
@@ -457,6 +462,21 @@ training_args = DPOConfig(
 
 These parameters only appear in the [published version](https://aclanthology.org/2025.tacl-1.22.pdf)
 
+### Statistical Rejection Sampling Improves Preference Optimization
+**📜 Paper**: https://huggingface.co/papers/2309.06657
+
+Proposes **RSO**, selecting stronger preference pairs via statistical rejection sampling to boost offline preference optimization; complements DPO/SLiC.
```
Comment (qgallouedec) [inline docs/source/paper_index.md]: same here, and for all the papers, a little code snippet would help

File: docs/source/paper_index.md [hunk 3]
```diff
@@ -509,6 +529,10 @@ SFTConfig(
     gradient_accumulation_steps=32,
 )
 ```
+### LoRA: Low-Rank Adaptation of Large Language Models
```
Comment (qgallouedec) [inline docs/source/paper_index.md:795]: LoRA isn't specific to SFT, So I'd rather have it in a dedicated section

File: docs/source/paper_index.md [hunk 4]
```diff
@@ -605,3 +634,33 @@ def add_margin(example):
...
+Foundational **PPO** objective with clipped ratios and minibatch epochs—baseline for many RL/RLHF variants used in TRL.
+
+### ZeRO: Memory Optimizations Toward Training Trillion-Parameter Models
+**📜 Paper**: https://huggingface.co/papers/1910.02054
+
+**ZeRO** partitions optimizer states/gradients/params to scale training efficiently; relevant when configuring DeepSpeed/Accelerate with TRL.
+
+
+[`GRPOTrainer`]: grpo_trainer
+[`CPOTrainer`]: cpo_trainer
```
Comment (qgallouedec) [inline docs/source/paper_index.md]: these can be removed

**Review by qgallouedec — COMMENTED** [review]
thanks for the contribution, I made a few comments, basically the idea is to align with the rest of the page

File: docs/source/paper_index.md (hunk 1 above)
Comment (SSusantAchary) [reply docs/source/paper_index.md]: moved the part to first

File: docs/source/paper_index.md (hunk 2 above)
Comment (SSusantAchary) [reply docs/source/paper_index.md]: added

File: docs/source/paper_index.md (hunk 3 above)
Comment (SSusantAchary) [reply docs/source/paper_index.md:795]: created section

File: docs/source/paper_index.md (hunk 4 above)
Comment (SSusantAchary) [reply docs/source/paper_index.md]: taken out

File: docs/source/paper_index.md [hunk 5]
```diff
@@ -5,8 +5,28 @@
...
+Introduces **GRPO** and shows strong math-reasoning gains from math-centric pretraining plus group-relative PPO-style optimization.  
+**Used in TRL via:** [`GRPOTrainer`]
+
+```python
+# Minimal GRPO setup (mirrors style used for other papers on the page).
+from trl import GRPOConfig
+
+training_args = GRPOConfig(
+    loss_type="grpo",
+    beta=0.0,                 # GRPO commonly trains without explicit KL in released configs
```
Comment (qgallouedec) [inline docs/source/paper_index.md]: In the original paper they don't use beta=0.0

File: docs/source/paper_index.md [hunk 6]
```diff
@@ -5,8 +5,28 @@
...
+    epsilon=2e-4,             # clip range (use paper/experiment settings if you mirror them)
+    epsilon_high=4e-4,        # upper clip (symmetrical if not specified)
```
Comment (qgallouedec) [inline docs/source/paper_index.md]: The goal here is to roughly match the setting used in the paper. So similarly to beta: in the original paper, they don't use asymmetric clipping

File: docs/source/paper_index.md [hunk 7]
```diff
@@ -232,10 +252,6 @@ trainer = PAPOTrainer(
 )
 ```
 
-## Direct Policy Optimization
-
-Papers relating to the [`DPOTrainer`]
-
```
Comment (qgallouedec) [inline docs/source/paper_index.md:422]: Can you please revert this change?

File: docs/source/paper_index.md [hunk 8]
```diff
@@ -457,6 +473,74 @@ training_args = DPOConfig(
...
+
+def rso_accept(ex):  # replace with your statistic (gap / z-score / judge score)
+    return ex.get("rso_keep", True)
+
+dpo_pairs = dpo_pairs.filter(rso_accept)
+
+model = AutoModelForCausalLM.from_pretrained("..."); tok = AutoTokenizer.from_pretrained("...")
+args = DPOConfig(loss_type="sigmoid", beta=0.1)
+trainer = DPOTrainer(model=model, args=args, tokenizer=tok, train_dataset=dpo_pairs)
+trainer.train()
```
Comment (qgallouedec) [inline docs/source/paper_index.md]:
```suggestion
from datasets import load_dataset
from trl import DPOConfig, DPOTrainer

train_dataset = load_dataset(...)

def rso_accept(example):  # replace with your statistic (gap / z-score / judge score)
    return example.get("rso_keep", True)

train_dataset = train_dataset.filter(rso_accept)

training_args = DPOConfig(loss_type="sigmoid", beta=0.1)
trainer = DPOTrainer(
    ...,
    args=training_args,
    train_dataset=train_dataset
)
trainer.train()
```

for consistency

File: docs/source/paper_index.md [hunk 9]
```diff
@@ -457,6 +473,74 @@ training_args = DPOConfig(
...
+from trl import DPOConfig, DPOTrainer
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+def rso_accept(ex):  # replace with your statistic (gap / z-score / judge score)
+    return ex.get("rso_keep", True)
+
+dpo_pairs = dpo_pairs.filter(rso_accept)
+
+model = AutoModelForCausalLM.from_pretrained("..."); tok = AutoTokenizer.from_pretrained("...")
+args = DPOConfig(loss_type="sigmoid", beta=0.1)
```
Comment (qgallouedec) [inline docs/source/paper_index.md]: isn't the loss supposed to be "hinged"?

File: docs/source/paper_index.md [hunk 10]
```diff
@@ -457,6 +473,74 @@ training_args = DPOConfig(
...
+for ex in prompts_ds:
+    out = policy.generate(**tok(ex["prompt"], return_tensors="pt").to(policy.device), max_new_tokens=256)
+    resp = tok.decode(out[0], skip_special_tokens=True)
+    score = ai_feedback_score(ex["prompt"], resp)
+    # build a (chosen, rejected) pair using score (e.g., compare vs baseline response)
+    new_pairs.append({"prompt": ex["prompt"], "chosen": resp, "rejected": ex["baseline"]})
+
+augmented_pairs = dpo_pairs.add_items(new_pairs)
+
+args = DPOConfig(loss_type="sigmoid", beta=0.1)
```
Comment (qgallouedec) [inline docs/source/paper_index.md]: there no need to pass the value if it matches the default. In other words, remove any occurence of like `loss_type="sigmoid"` or  `beta=0.1`

File: docs/source/paper_index.md [hunk 11]
```diff
@@ -508,6 +592,50 @@ SFTConfig(
...
+
+trainer = SFTTrainer(
+    model=model,
+    args=args,
+    tokenizer=tok,
+    peft_config=peft_cfg,   # <- LoRA enabled
+    train_dataset=...,
+)
+trainer.train()
+
```
Comment (qgallouedec) [inline docs/source/paper_index.md]:
```suggestion
from peft import LoraConfig
from trl import SFTTrainer

trainer = SFTTrainer(
    ...,
    peft_config=LoraConfig(),
)
```

the more minimal, the clearer

File: docs/source/paper_index.md (hunk 5 above)
Comment (SSusantAchary) [reply docs/source/paper_index.md]: removed

File: docs/source/paper_index.md (hunk 6 above)
Comment (SSusantAchary) [reply docs/source/paper_index.md]: cleaned the above

File: docs/source/paper_index.md (hunk 7 above)
Comment (SSusantAchary) [reply docs/source/paper_index.md:422]: reverted

File: docs/source/paper_index.md (hunk 9 above)
Comment (SSusantAchary) [reply docs/source/paper_index.md]: updated

File: docs/source/paper_index.md (hunk 8 above)
Comment (SSusantAchary) [reply docs/source/paper_index.md]: minimal

File: docs/source/paper_index.md (hunk 11 above)
Comment (SSusantAchary) [reply docs/source/paper_index.md]: updated

**Comment by SSusantAchary** [comment]
@qgallouedec , pls review the changes

**Comment by SSusantAchary** [comment]
@qgallouedec , would it be possible to get some time and review the updates , TY

**Comment by qgallouedec** [comment]
Thanks!
I removed a few section, because the code or explanation were just not correct (example: PRM paper using a code example with `GRPOTrainer` but not the `PRMTrainer`). These sections were most likely LLM-generated?
I'm just having a final pass and we're good to merge
